cdap-app-fabric/src/main/java/io/cdap/cdap/security/auth/AuditLogSubscriberService.java

```diff
@@ -0,0 +1,167 @@
+import io.cdap.cdap.spi.data.StructuredTableContext;
+import io.cdap.cdap.spi.data.transaction.TransactionRunner;
+import org.apache.tephra.TxConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayDeque;
+import java.util.Iterator;
+import java.util.Queue;
+import javax.annotation.Nullable;
+
+/**
+ * This class subscribes to the audit log messaging topic, reads and processes the messages into {@link AuditLogContext}
+ * and delegates the batch of AuditLogContexts to External Auth service using {@link AuditLoggerSpi}, which would
+ * further publish as configured.
+ */
+public class AuditLogSubscriberService extends AbstractMessagingSubscriberService<AuditLogContext> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AuditLogSubscriberService.class);
+  private static final Gson GSON = new Gson();
+
+  private final MultiThreadMessagingContext messagingContext;
+  private final TransactionRunner transactionRunner;
+  private final AccessControllerInstantiator accessControllerInstantiator;
+
+  @Inject
+  AuditLogSubscriberService(CConfiguration cConf, MessagingService messagingService,
+                            MetricsCollectionService metricsCollectionService,
+                            TransactionRunner transactionRunner,
+                            AccessControllerInstantiator accessControllerInstantiator) {
+
+    super(
+      NamespaceId.SYSTEM.topic(cConf.get(Constants.AuditLogging.AUDIT_LOG_EVENT_TOPIC)),
+      cConf.getInt(Constants.AuditLogging.AUDIT_LOG_FETCH_SIZE),
+      cConf.getInt(TxConstants.Manager.CFG_TX_TIMEOUT),
+      cConf.getInt(Constants.AuditLogging.AUDIT_LOG_POLL_DELAY_MILLIS),
+      RetryStrategies.fromConfiguration(cConf, Constants.AuditLogging.AUDIT_LOG_WRITER_RETRY_PREFIX),
+      metricsCollectionService.getContext(ImmutableMap.of(
+        Constants.Metrics.Tag.COMPONENT, Constants.Service.MASTER_SERVICES,
+        Constants.Metrics.Tag.INSTANCE_ID, "0",
+        Constants.Metrics.Tag.NAMESPACE, NamespaceId.SYSTEM.getNamespace(),
+        Constants.Metrics.Tag.TOPIC, cConf.get(Constants.AuditLogging.AUDIT_LOG_EVENT_TOPIC),
+        Constants.Metrics.Tag.CONSUMER, Constants.AuditLogging.AUDIT_LOG_CONSUMER_WRITER_SUBSCRIBER
+      )));
+    this.messagingContext = new MultiThreadMessagingContext(messagingService);
+    this.transactionRunner = transactionRunner;
+    this.accessControllerInstantiator = accessControllerInstantiator;
+  }
+
+  /**
+   * Returns the {@link TransactionRunner} for executing tasks in transaction.
+   */
+  @Override
+  protected TransactionRunner getTransactionRunner() {
+    return transactionRunner;
+  }
+
+  /**
+   * Loads last persisted message id within a transaction for the topic of AUDIT LOG EVENTS.
+   * The returned message id will be used as the starting message id (exclusive) for the first fetch.
+   */
+  @Nullable
+  @Override
+  protected String loadMessageId(StructuredTableContext context) throws Exception {
```

Comment:
nit: Huh, I think at some point we would want to have an Abstract class for this boilerplate code

Reply:
Yeah true, Created jira for this : https://cdap.atlassian.net/browse/CDAP-21077 . 

Will try to take it up in a Separate PR.